src/Control/Controller.php

```diff
@@ -690,14 +697,31 @@ public static function join_links($arg = null)
             }
             if ((is_string($arg) && $arg) || is_numeric($arg)) {
                 $arg = (string) $arg;
-                if ($result && substr($result ?? '', -1) != '/' && $arg[0] != '/') {
-                    $result .= "/$arg";
+                if (strlen($result) == 0) {
+                    $result .= $arg;
+                } elseif (substr($result, -1) == '/') {
+                    $result .= trim($arg, '/');
                 } else {
-                    $result .= (substr($result ?? '', -1) == '/' && $arg[0] == '/') ? ltrim($arg, '/') : $arg;
+                    $result .= '/' . trim($arg, '/');
                 }
             }
         }
 
+        // add trailing slash if enabled and url does not contain a file extension
+        if (static::config()->get('add_trailing_slash') === true
+            && substr($result, -1) !== '/'
+            && !preg_match('/^(.*)\.([A-Za-z][A-Za-z0-9]*)$/', $result)
+        ) {
+            $result = rtrim($result, '/') . '/';
+        }
+
+        // add trailing slash to root urls (relative or absolute)
+        if (substr($result, -1) !== '/'
+            && (strlen(rtrim($result, '/')) == 0 || !parse_url($result, PHP_URL_PATH))
+        ) {
+            $result = rtrim($result, '/') . '/';
```

Comment:
Is the `rtrim` necessary here? I thought that's what the `substr($result, -1) !== '/'` check was for. The check itself seems to have a bit of double handling, as well.

What exactly does this do, anyway? It seems like this whole check is just "if there's no path, make sure the url ends with a trailing slash". Is that the correct and desired behaviour even when the config says to _not_ add trailing slashes?

Reply:
You're probably right. I have tried to simplify this, but whenever I change something here, the graphql calls in the CMS fail. I have no idea about graphql, so I can't fix it on that side.